lib/timers.js

```diff
@@ -62,55 +118,81 @@ function listOnTimeout() {
   var now = Timer.now();
   debug('now: %s', now);
 
-  var diff, first, threw;
-  while (first = L.peek(list)) {
-    diff = now - first._idleStart;
+  var diff, timer, threw;
+  while (timer = L.peek(list)) {
+    diff = now - timer._idleStart;
+
+    // Check if this loop iteration is too early for the timer.
+    // This happens if one of two things is true:
+    // - There are more timers scheduled for later in the list. (Most common)
+    // - The earliest timer was canceled (unenrolled). (Less common)
     if (diff < msecs) {
       list.start(msecs - diff, 0);
       debug('%d list wait because diff is %d', msecs, diff);
       return;
-    } else {
-      L.remove(first);
-      assert(first !== L.peek(list));
+    }
 
-      if (!first._onTimeout) continue;
+    // The actual logic for when a timeout happens.
 
-      // v0.4 compatibility: if the timer callback throws and the
+    L.remove(timer);
+    assert(timer !== L.peek(list));
+
+    if (!timer._onTimeout) continue;
+
+    var domain = timer.domain;
+    if (domain) {
+
+      // If the timer callback throws and the
       // domain or uncaughtException handler ignore the exception,
       // other timers that expire on this tick should still run.
       //
       // https://github.com/joyent/node/issues/2631
-      var domain = first.domain;
-      if (domain && domain._disposed)
+      if (domain._disposed)
         continue;
```

Comment:
Do you mean the `return` statement in [` _makeTimerTimeout`](https://github.com/nodejs/node/pull/4007/files#diff-0a5d4868b2b9b17cf9e2c11f1bd1311eL509)? It was `return` statement for unrefed timers because that logic is encapsulated in a function, so that `return` is the equivalent to the `continue` in the `listOnTimeout` loop.

There's [at least one place where we used to have a test for this behavior](https://github.com/nodejs/node/pull/3990/files#diff-11b55dc908c51630a9463889515844bbL29) for timers created with `setTimeout`, but I removed it with https://github.com/nodejs/node/pull/3990.

Basically, removing https://github.com/nodejs/node/pull/3990 was a mistake on my side, because I assumed that it was testing that `process.domain` was cleared properly when an error was thrown from within a timer, when it was actually testing that a timer attached to a disposed domain wouldn't run. It's obvious to me in hindsight, I apologize for the confusion.

The test that https://github.com/nodejs/node/pull/3990 added is good though. So what I'll do is that I'll put back the original test that https://github.com/nodejs/node/pull/3990 removed, but with more comments to explain clearly what it tests, and move the test that https://github.com/nodejs/node/pull/3990 added to a new file.

Then, we can add a test equivalent to the test that https://github.com/nodejs/node/pull/3990 removed, but for unrefed timers, in a new file too.